update project

tansunasyrova opened the pull request.

tansunasyrova committed:
hhh

tansunasyrova committed:
new update

tansunasyrova commented:
new update 13/04

tansunasyrova committed:
new update (test.py was added)

Pandylandy commented on synthesis.py:
```diff
@@ -1 +1,148 @@
+        return self.state, reward, done, info
+
+    def reset(self):
+        with db_session:
```
тут нужна сессия?

Pandylandy commented on synthesis.py:
```diff
@@ -1 +1,148 @@
+
+    single_rules = load(open('/home/tansu/rnn/indexed_single_rules.pickle', 'rb'))
+    double_rules = load(open('/home/tansu/rnn/indexed_two_rules.pickle', 'rb'))
+    groups = load(open('/home/tansu/rnn/groups_common_very_new.pickle', 'rb'))
```
пути тоже можно вынести в конфиг

Pandylandy commented on helper.py:
```diff
@@ -1,5 +1,69 @@
+        molecule = structure
+        for i, fg in self.groups.items():
+            if fg < molecule:
+                group_list.append(i)
```
сделай через списковое выражение. 4 строчки -> 1 строчка. быстрее будет работать.

Pandylandy commented on helper.py:
```diff
@@ -1,5 +1,69 @@
+
+
+def reactions_by_fg(self, group_id_list, n):
+    reactions_list = []
```
тут сделай без n. используй keyword аргументы функции.

tansunasyrova replied:
не очень поняла

Pandylandy replied:
сделай аргумент по умолчанию. типа однокомпонентная это тру. посмотри в инете кейворд аргументы

tansunasyrova replied:
все, разобралась

Pandylandy commented on helper.py:
```diff
@@ -1,5 +1,69 @@
+        for react in self.single_rules:
+            reaction_groups = self.single_rules[react]
+            if any(group in reaction_groups for group in group_id_list):
+                reactions_list.append(react)
```
чтобы этот код не повторялся, видишь, он почти одинаковый

tansunasyrova committed:
я сделала, чтобы был путь синтеза

Pandylandy reviewed (commented):
и сделай везде импорты по pep8

tansunasyrova committed:
исправила замечаная
только не пойму, почему функция group_list подчеркивается красным на 106, 119, 128 строчках (в файле synthesis)

Pandylandy commented on helper.py:
```diff
@@ -1,6 +1,10 @@
 from CIMtools.preprocessing import FragmentorFingerprint
+from config import *
+
+db = load_schema('bb', user=user, password=password, database=database, host=host)
```
подключаться к базе можно один раз. поэтому в этом файле не надо, ты уже в главном создала

Pandylandy commented on helper.py:
```diff
@@ -28,42 +32,35 @@ def evaluation(mol1, mol2):
-                group_list.append(i)
+        with db_session:
+            group_list = [i for i, fg in groups.items() if fg < molecule]
     return group_list
```
у меня тут другие номера строк. запусти код, поймешь в чем ошибка. но если подчеркивает красным - значит, очевидная ошибка. разберись. и всегда делай коммит, когда не будет таких ошибок. ничего не должно подчеркивать красным

Pandylandy commented on prepare.py:
```diff
@@ -1,15 +1,16 @@
+from CGRtools.files import RDFread, SDFwrite
 import pickle
 from datetime import datetime
+from config import *
```
импорты везде по pep8 сделай

Pandylandy commented on prepare.py:
```diff
@@ -1,15 +1,16 @@
 
-db = load_schema('bb')
+db = load_schema('bb', user=user, password=password, database=database, host=host)
 Reagents = db.Molecule
```
это убери

Pandylandy commented on synthesis.py:
```diff
@@ -4,12 +4,6 @@
 from CGRtools import Reactor
 from CGRtools.containers import MoleculeContainer
 from CGRtools.containers import ReactionContainer
```
pep8

Pandylandy commented on synthesis.py:
```diff
@@ -36,15 +30,15 @@ class SimpleSynthesis(gym.Env):
-    double_rules = load(open('/home/tansu/rnn/indexed_two_rules.pickle', 'rb'))
-    groups = load(open('/home/tansu/rnn/groups_common_very_new.pickle', 'rb'))
+    single_rules = single_rules
+    double_rules = double_rules
```
какой смысл?

Pandylandy replied:
зачем плодить ссылки

Pandylandy commented on helper.py:
```diff
@@ -28,42 +32,35 @@ def evaluation(mol1, mol2):
 
 
-__all__ = ['dictionary', 'evaluation', 'best_n_molecules', 'reactions_by_fg', 'group_list']
+__all__ = [dictionary, evaluation, best_n_molecules, reactions_by_fg, group_list]
```
?

Pandylandy reviewed (commented):
сначала всегда тестишь, если работает - заливаешь)

Pandylandy commented on prepare.py:
```diff
@@ -1 +1,209 @@
+
+def mol_number():
+    return [mol.id for mol in db.Molecule.select()]
+
```
это вроде договорились выкинуть

Pandylandy commented on prepare.py:
```diff
@@ -1 +1,209 @@
+        mol_dict[n] = id
+        a += 1
+    mol_dict[a] = None
+    mol_dict[a+1] =
```
что это?

Pandylandy commented on prepare.py:
```diff
@@ -1 +1,209 @@
+    mol1_fp = FingerprintMolecule.get_fingerprint(mol1)
+    mol2_fp = FingerprintMolecule.get_fingerprint(mol2)
+    mol1_c, mol2_c, common = len(mol1_fp), len(mol2_fp), len(mol1_fp.intersection(mol2_fp))
+    return common / (mol1_c + mol2_c - common)
```
почему у тебя функции из хелпера тут дублируются? в этом файле ж подготовка базы только должна быть

tansunasyrova committed:
исправила еще некоторые ошибки, но теперь он не видит базу данных

tansunasyrova committed:
я вроде исправила, но у меня все равно не работает. не знаю, что делать

Pandylandy commented on helper.py:
```diff
@@ -1,10 +1,10 @@
 from pony.orm import db_session
 from operator import itemgetter
-from CGRdb import load_schema
+from CGRdb import load_schema, Molecule
```
это так не работает. и сделай уже нормальные импорты!

Pandylandy replied:
ты импортируешь класс

Pandylandy commented on helper.py:
```diff
@@ -52,15 +52,15 @@ def reactions_by_fg(group_id_list, single=True):
     if molecule:
         with db_session:
-            group_list = [x.id for x in db.Molecule[molecule.id].classes]
+            groups_list = [x.id for x in Molecule[molecule.id].classes]
```
а тут хочешь у класса неизвестно что получить по id

Pandylandy reviewed (commented):
ты не все замечания исправила. внимательно посмотри весь conversation. + база данных грузится ОДИН раз в главном файле

tansunasyrova committed:
Протестировала код несколько раз, всё должно работать)
В файле "test" функция random работает, путь в конце выдает. Но я пока закомментила ее, потому что сама прописала алгоритм действий со всеми необычными случаями (типа next и none) - я прокомментировала там каждый шаг, чтобы тебе было удобнее понимать, что происходит. Путь всегда выдается правильный (даже если мы просим next несколько раз подряд).

Еще каждый раз, когда получаем список из 10 реакций и сортируем его в порядке убывания танимото, этот список распечатывается (чтобы удобнее было смотреть)

Pandylandy commented on synthesis.py:
```diff
@@ -100,39 +102,49 @@ def add_reagent(self, action):
+                rules = reactions_by_fg(groups_list)
                 for rule in rules:
                     reactor = Reactor(rule, delete_atoms=True)
                     reactions = list(reactor([self.state]))
```
Если ты используешь только 1 реакцию потом, то возьми просто next от итератора

Pandylandy commented on synthesis.py:
```diff
@@ -100,39 +102,49 @@ def add_reagent(self, action):
+                # print('RULES', rules)
                 for rule in rules:
                     reactor = Reactor(rule, delete_atoms=True)
                     reactions = list(reactor([self.state, reagent]))
```
то же самое

Pandylandy commented on synthesis.py:
```diff
@@ -85,7 +83,11 @@ def add_reagent(self, action):
                     return self.state, None, reward, {'info': 'no another reaction products at the list'}
             if len(reactions_list) > 1:
```
Если это проверка на существование продуктов, то не обязательно смотреть длину. Не пустой список это True, пустой - False

Pandylandy commented on synthesis.py:
```diff
@@ -85,7 +83,11 @@ def add_reagent(self, action):
+                    path[-1] = reaction
+                else:
+                    path.append(reaction)
+                path[-1] = reaction  # заменяем последнюю реакцию в пути
```
не очень понятно. Если в пат ничего нет - добавляешь реакцию, а потом ее же заменяешь. много действий)

Pandylandy commented on synthesis.py:
```diff
@@ -85,7 +83,11 @@ def add_reagent(self, action):
+                    path.append(reaction)
+                path[-1] = reaction  # заменяем последнюю реакцию в пути
                 self.reactions_list = reactions_list
                 self.path = path
```
работай сразу с self.path

Pandylandy commented on synthesis.py:
```diff
@@ -85,7 +83,11 @@ def add_reagent(self, action):
                     return self.state, None, reward, {'info': 'no another reaction products at the list'}
             if len(reactions_list) > 1:
                 state, reaction, reward = reactions_list.pop(0)
```
а зачем хранить молекулу и реакцию, если в реакции уже есть эта молекула? в этом списке у нас только реакция и танимото. и даже танимото можно засунуть в метаданные реакции

Pandylandy commented on synthesis.py:
```diff
@@ -141,8 +153,11 @@ def add_reagent(self, action):
-                    path.append(reaction) # реакция добавляется в путь, если она не была добавлена только что
+                if len(path) >= 1:
+                    if path[-1] != reaction:
+                        path.append(reaction)  # реакция добавляется в путь, если она не была добавлена только что
```
только что - это как? это долгая проверка. сильно будет замедлять. надо обойтись без нее , и пока мне непонятно, зачем она нужна

Pandylandy commented on synthesis.py:
```diff
@@ -141,8 +153,11 @@ def add_reagent(self, action):
+                    path.append(reaction)
                 self.reactions_list = reactions_list
                 self.path = path
                 return state, reaction, reward, {'info': 'the last molecule at the list'}
```
нам нужна реакция в step? тут только стейт и реворд должен быть. вся инфа о реакциях в path, кому надо - смотрит там

Pandylandy commented on tests.py:
```diff
@@ -1,44 +1,59 @@
+e = play.step(736785)  # none - однореагентная реакция 2+1
+print('5!', e)
+f = play.step(736784)  # 2-реагентная реакция текущей молекулы с последней молекулой "736784" 3+1
+print('6!', f)  # получилось 2 новые реакции, почему-то ReactionContainer у них разные, но сами реакции одинаковые
```
в каком смысле разные? картинки хочу)

Pandylandy commented on tests.py:
```diff
@@ -1,44 +1,59 @@
+j = play.step(736785)  # none - однореагентная реакция 4+1
+print('9!', j)
+k = play.step(736786)  # next - просим следующую молекулу 4+1(!), но получаем такую же молекулу, =>
+#  тк там одинаковые реакции (ПОЧEМУ?)
```
у тебя нет проверки на уже встретившуюся реакцию в пределах одного шага. поэтому там одинаковые реакции. надо добавить

Pandylandy commented on tests.py:
```diff
@@ -1,44 +1,59 @@
+print('PATH:', play.render())  # просим путь, в котором должно быть 5 реакций, он верный
+m = play.step(736786)  # next - просим следующую молекулу 4+1(!!)
+print('11!', m)  # next - просим следующую молекулу 4+1(!!)
+print('PATH:', play.render())  # просим путь, в котором должно быть 5 реакций, он верный
```
нужны картинки.
1. просто 5 шагов
2. шаги с Next на разных этапах

Pandylandy reviewed (commented):
Во-первых, покажи картинки с реакциями, во-вторых, если next выдает одинаковые молекулы - это значит, что не все работает. Path нужно еще проработать. и посмотри issue, я там отметила тебя

tansunasyrova committed:
замечания исправила, единственное, он теперь не добавляет в путь одну реакцию (строчка 42 в test)
пока я так и не поняла, почему(

- next теперь возвращает самый первый продукт, который был до next (когда продукты в списке заканчиваются)
- список из 10 лучших полекул теперь состоит из уникальных молекул и реакций (исправила это в функции best_n_molecules в файле helper)
- остальное тоже поправила, кроме того что reaction list по прежнему хранит  state, reaction, reward, потому что не понимаю, как можно по друогму сделать, мне кажется, удобнее всего, что он возвращает их по отдельности

tansunasyrova committed:
теперь должно работать

tansunasyrova committed:
теперь должно работать

Pandylandy commented on helper.py:
```diff
@@ -35,8 +35,8 @@ def best_n_molecules(reactions_list, target, n):
         for product in reaction.products:
-            # print('PRODUCT', product)
             tanimoto_list.append((product, reaction, evaluation(product, target)))
+    tanimoto_list = list(set(tanimoto_list))
```
надо отсекать дубликаты на этапе генерации. в функции add reagent.

Pandylandy commented on synthesis.py:
```diff
@@ -72,31 +73,30 @@ def reset(self):
+                        self.path[-1] = reaction  # заменяем последнюю реакцию в пути
+                    else:
+                        self.path.append(reaction)
+                    self.reactions_list[0] = self.first_step
```
ты заменяешь первый элемент, то есть тот, который там был - исчезнет. так не надо) ты теряешь информацию

Pandylandy commented on synthesis.py:
```diff
@@ -122,12 +122,12 @@ def add_reagent(self, action):
                 for rule in rules:
                     reactor = Reactor(rule, delete_atoms=True)
-                    reactions = list(reactor([self.state, reagent]))
+                    reactions = next(reactor([self.state, reagent]), None)
```
тогда это не реакции, а одна реакция. То есть reaction. ну это просто замечание

Pandylandy commented on synthesis.py:
```diff
@@ -144,22 +144,20 @@ def add_reagent(self, action):
             reactions_list = best_n_molecules(reactions_list, self.target, 10)
-            print('REACT LIST 10 best', reactions_list)
+            print('REACT LIST 10 best', (len(reactions_list)), reactions_list)
+            self.first_step = reactions_list[0]
```
идея рабочая, но не самая лучшая. попробуй просто добавить начальный продукт в конец. тогда будет легко его взять в next

Pandylandy commented on helper.py:
```diff
@@ -34,10 +35,12 @@ def best_n_molecules(reactions_list, target, n):
         # print('!reaction.products!', reaction.products)
-        for product in reaction.products:
-            tanimoto_list.append((product, reaction, evaluation(product, target)))
+        product = [sorted(reaction.products, key=lambda x: x.molecular_mass, reverse=True)][0]
```
это делаешь в add reagent. лишние реакции не храним. сюда попадают уже готовые реакции с одним продуктом и просто сортируются. и тут не по массе, а по длине, я же написала. масса может быть высокая у маленькой молекулки). достаточно просто использовать max от списка длин. без sorted

Pandylandy commented on helper.py:
```diff
@@ -34,10 +35,12 @@ def best_n_molecules(reactions_list, target, n):
-        for product in reaction.products:
-            tanimoto_list.append((product, reaction, evaluation(product, target)))
+        product = [sorted(reaction.products, key=lambda x: x.molecular_mass, reverse=True)][0]
+        meta = {'tanimoto': evaluation(product[0], target)}
```
добавь сюда еще rule. по которому прошла реакция. и можно сразу в мету передавать словарь, без переменной

Pandylandy commented on synthesis.py:
```diff
@@ -76,30 +75,34 @@ def add_reagent(self, action):
                 else:
-                    state, reaction, reward = self.reactions_list.pop(0)
+                    reaction = self.reactions_list.pop(0)
+                    state = reaction.products[0]
```
можно без присваивания переменных. занимает время. нам дорога каждая секунда) потому что сейчас это очень долго работает скорее всего. запусти на 500-1000 шагов, оцени среднее время одного шага

Pandylandy commented on tests.py:
```diff
@@ -39,7 +39,7 @@ def random_func(n):
 d = play.step(5)  # 2-реагентная реакция текущей молекулы с молекулой "5", но она не получилась, поэтому, state остается
 print('4!', d)
-e = play.step(736785)  # none - однореагентная реакция 2+1 ОН НЕ ДОБАВИЛ ЕЕ В ПУТЬ
+e = play.step(736785)  # none - однореагентная реакция 2+1
```
называй коммиты так, чтобы можно было  понять, что изменилось. и как изменилось. два коммита одинаковых , причем в последнем это просто комментарий - ну не очень вообще

tansunasyrova committed:
делаю этот комит, чтобы были видны дальнейшие изменения (пришлось весь код заново вставить)

+добавила файл search в котором я делаю монте-карло

tansunasyrova committed:
исправила твои последние замечания

Pandylandy commented on synthesis.py:
```diff
@@ -125,30 +125,37 @@ def add_reagent(self, action):
+            react_list = []
+            for reaction in reactions_list:
+                product = max(reaction.products)
+                meta = {'tanimoto': evaluation(product, self.target)}
```
а правило где?

tansunasyrova replied:
про правило забыла( сейчас сделаю

Pandylandy commented on synthesis.py:
```diff
@@ -41,7 +41,7 @@ def __init__(self, target, step_number=10):
         self.max_steps = step_number
         self.np_random = None
-        self.first_step = None
+        self.saved_reactions = []
```
это для чего?

tansunasyrova replied:
ты же говорила, что если next доедет до конца списка, чтобы он начал заново выдавать молекулы с начала списка
дело в том, что next делает pop, поэтому эти молекулы теряются. поэтому я сохраняю вот так

Pandylandy replied:
это странно. сделай как я написала до этого и просто добавь доп. проверку. 
ты же говорила, что если next доедет до конца списка, чтобы он начал заново выдавать молекулы с начала списка - такого я не говорила)

Pandylandy replied:
надо сделать так, чтобы когда продукты закончились - он выдал молекулу , которая была стейтом до next

Pandylandy commented on synthesis.py:
```diff
@@ -125,30 +125,37 @@ def add_reagent(self, action):
+            reactions_list = list(set(reactions_list))
+            react_list = []
+            for reaction in reactions_list:
+                product = max(reaction.products)
```
так не сработает. надо было длину брать

tansunasyrova replied:
а он как раз так выбирает самую длинную молекулу
я в консоли просто проверяла)
если делать max(len(продукт)), то он выдает int - длину самого длинного продукта

Pandylandy commented on search.py:
```diff
@@ -0,0 +1,61 @@
+import random
+import os
+import math
+from mcts import mcts
```
а это что?

tansunasyrova replied:
mcts осталось с неудачной попытки использовать чужой монте-карло, я удалю это

Pandylandy commented on .idea/remote-mappings.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
эти файлы не надо сюда добавлять

tansunasyrova committed:
- в meta у реакций добавила rule
- увеличила словарь с молекулами - тепрь там однореагентных реакций (команд none) на 20% больше
- теперь первая молекула не вступает в однореагентную реакцию, а начинает реагровать со следующей (как сказал ТИ)
- в search сделала функцию vector - алгоритм монте-карло с векторами молекул
- еще есть функция simulated_random - она изначально выбирает молекулу с танимото не меньше 0.3 и работает рандомно. пока не уверена, что я до конца там все правильно сделала

tansunasyrova committed:
synthesis_logP и synthesis_pharm - это модификации среды под поиск молекулы с максимальной липофильностью и фармакофорным сходством
Там небольшие изменения, но сделала их в отдельные файлы, чтобы не путаться)
